### PR TITLE
Implemented lmul evaluator

### PR DESCRIPTION
**Pull request by hsoontie**
Implemented lmul evaluator for AArch64

### CODE REVIEW

File: compiler/aarch64/codegen/BinaryEvaluator.cpp [hunk 1]
```diff
@@ -159,10 +184,37 @@ OMR::ARM64::TreeEvaluator::imulhEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
+      {
+      int64_t value = secondChild->getLongInt();
+      if (value > 0 && cg->convertMultiplyToShift(node))
+         {
+         trgReg = cg->evaluate(node);
+         return trgReg;
+         }
+      else
+         {
+         trgReg = cg->allocateRegister();
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/BinaryEvaluator.cpp]: You may not need to allocate a new register for the target if one of the source children dies at this instruction.  If the reference count on either source child is 1 (and it's evaluated into a register) then you should re-use it.

File: compiler/aarch64/codegen/BinaryEvaluator.cpp [hunk 2]
```diff
@@ -159,10 +184,37 @@ OMR::ARM64::TreeEvaluator::imulhEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
+         mulConstant64(node, trgReg, src1Reg, value, cg);
+         }
+      }
+   else
+      {
+      TR::Register *src2Reg = cg->evaluate(secondChild);
+      trgReg = cg->allocateRegister();
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/BinaryEvaluator.cpp]: Same comment about the target register above.

**Comment by hsoontie** [comment]
@0xdaryl Here is an update! It's all together now as well.

File: compiler/aarch64/codegen/BinaryEvaluator.cpp [hunk 3]
```diff
@@ -160,9 +184,48 @@ OMR::ARM64::TreeEvaluator::imulhEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
+      {
+      trgReg = cg->allocateRegister();
+      }
+
+   if (secondChild->getOpCode().isLoadConst() && secondChild->getRegister() == NULL)
+      {
+      int64_t value = secondChild->getLongInt();
+      if (value > 0 && cg->convertMultiplyToShift(node))
+         {
+         trgReg = cg->evaluate(node);
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/BinaryEvaluator.cpp]: I think it is possible to reach here by taking the path through L203, which allocates a target register.  If that's the case, then that register will be lost.  Is it possible to move this logic where you decide to convert a multiply to a shift before the initial analysis starting at L193?  Alternatively, can you avoid allocating a target register if you know you'll be coming down this path?

File: compiler/aarch64/codegen/BinaryEvaluator.cpp [hunk 4]
```diff
@@ -160,9 +184,48 @@ OMR::ARM64::TreeEvaluator::imulhEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
+   TR::Register *src2Reg = NULL;
+   TR::Register *trgReg = NULL;
+
+   if(1 == firstChild->getReferenceCount())
+      {
+      trgReg = src1Reg;
+      }
+   else if(1 == secondChild->getReferenceCount() && secondChild->getRegister() != NULL)
+      {
+      trgReg = src2Reg;
```
[the remark this answers is not in the document]
Comment (hsoontie) [reply compiler/aarch64/codegen/BinaryEvaluator.cpp]: Thank you, I believe you're right, I think I'd meant to set it as I made the check.

**Comment by hsoontie** [comment]
@0xdaryl 
Here is a small update. I moved the part of the logic you mentioned upwards.

**Comment by hsoontie** [comment]
@0xdaryl The whitespace should be fixed up now! I will look into creating an issue for imul.

**Comment by 0xdaryl** [comment]
@genie-omr build aarch64

**Pull request merged**